mount/losetup_linux.go

```diff
@@ -0,0 +1,229 @@
+	}
+	return uint32(num), nil
+}
+
+func setupLoopDev(backingFile, loopDev string, param LoopParams) (devFile *os.File, err error) {
+	// 1. Open backing file and loop device
+	oflags := os.O_RDWR
+	if param.Readonly {
+		oflags = os.O_RDONLY
+	}
+	back, err := os.OpenFile(backingFile, oflags, 0)
+	if err != nil {
+		return nil, errors.Errorf("could not open backing file: %v", err)
+	}
+	defer back.Close()
+
+	loopFile, err := os.OpenFile(loopDev, oflags, 0)
+	if err != nil {
+		return nil, errors.Errorf("could not open loop device: %v", err)
+	}
+	defer func() {
+		if err != nil {
+			loopFile.Close()
+		}
+	}()
+
+	// 2. Set FD
+	if _, _, err = ioctl(loopFile.Fd(), ioctlSetFd, back.Fd()); err != nil {
+		return nil, errors.Errorf("could not set loop fd: %v", err)
+	}
+
+	// 3. Set Info
+	info := loopInfo{}
+	copy(info.fileName[:], []byte(backingFile))
+	if param.Readonly {
+		info.flags |= loFlagsReadonly
+	}
+	if param.Autoclear {
+		info.flags |= loFlagsAutoclear
+	}
+	if param.Direct {
+		info.flags |= loFlagsAutoclear
+	}
+	if _, _, err := ioctl(loopFile.Fd(), ioctlSetStatus64, uintptr(unsafe.Pointer(&info))); err != nil {
+		// Retry w/o direct IO flag in case kernel does not support it. The downside is that
+		// it will suffer from double cache problem.
+		info.flags &= ^(uint32(loFlagsDirectIO))
+		if _, _, err := ioctl(loopFile.Fd(), ioctlSetStatus64, uintptr(unsafe.Pointer(&info))); err != nil {
+			ioctl(loopFile.Fd(), ioctlClrFd, 0)
+			return nil, errors.Errorf("cannot set loop info:%v", err)
+		}
+	}
+
+	return loopFile, nil
+}
+
+// setupLoop looks for (and possibly creates) a free loop device, and
+// then attaches backingFile to it.
+//
+// When autoclear is true, caller should take care to close it when
+// done with the loop device. The loop device file handle keeps
+// loFlagsAutoclear in effect and we rely on it to clean up the loop
+// device. If caller closes the file handle after mounting the device,
+// kernel will clear the loop device after it is umounted. Otherwise
+// the loop device is cleared when the file handle is closed.
+//
+// When autoclear is false, caller should be responsible to remove
+// the loop device when done with it.
+//
+// Upon success, the file handle to the loop device is returned.
+func setupLoop(backingFile string, param LoopParams) (*os.File, error) {
+	var loopDev string
+
+	for retry := 1; retry < 10000; retry++ {
+		num, err := getFreeLoopDev()
+		if err != nil {
+			return nil, err
+		}
+
+		loopDev = fmt.Sprintf(loopDevFormat, num)
+		loopFile, err := setupLoopDev(backingFile, loopDev, param)
+		if err != nil {
+			// Per util-linux/sys-utils/losetup.c:create_loop(),
+			// free loop device can race and we end up failing
+			// with EBUSY when trying to set it up.
+			if strings.Contains(err.Error(), ebusyString) {
+				// Fallback a bit to avoid live lock
+				time.Sleep(time.Millisecond * time.Duration(rand.Intn(retry*10)))
```

Comment:
If I get this logic correctly, this func may sleep way too long:
```go
	total := time.Duration(0)
	for retry := 1; retry < 10000; retry++ {
                // time.Sleep(...)
		total += time.Millisecond * time.Duration(rand.Intn(retry*10))
	}
	fmt.Printf("%f", total.Hours()) // Prints 70 hours.
```

Reply:
I was trying to follow the logic in `losetup.c` (https://github.com/karelzak/util-linux/blob/master/sys-utils/losetup.c#L548) where it retries forever on `EBUSY`. @AkihiroSuda suggested using a max retry count so I added this extremely large retry.

Reply:
I'd propose some reasonable sleep times here. Like if we can't do it within 1 min, then just fail. @AkihiroSuda WDYT?

Reply:
nvm. I've changed it to retry at most 200 times that can translate into a little more than 1 minute.